### PR TITLE
Implement HTTP/2 input flow control

### PR DESCRIPTION
**Pull request by halter73**
#2710

### CODE REVIEW

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs [hunk 1]
```diff
@@ -188,6 +215,21 @@ private void AbortCore(ConnectionAbortedException abortReason)
 
             // Unblock the request body.
             RequestBodyPipe.Writer.Complete(new IOException(CoreStrings.Http2StreamAborted, abortReason));
+
+            // Stop counting any buffered bytes for this stream towards the connection input flow-control window.
+            _inputFlowControl.Abort();
```
Comment (Tratcher) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: This needs to update the connection flow control.

Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: It does internally.

File: src/Kestrel.Core/Internal/Http2/FlowControl/Http2StreamInputFlowControl.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            {
+                // Stream-level flow control was aborted. Any unread bytes have already been returned to the connection
+                // flow-control window by Abort().
+                return;
+            }
+
+            if (streamWindowUpdateSize > 0)
+            {
+                // Writing with the FrameWriter should only fail if given a canceled token, so just fire and forget.
+                _ = _frameWriter.WriteWindowUpdateAsync(_streamId, streamWindowUpdateSize);
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/FlowControl/Http2StreamInputFlowControl.cs]: Stream and connection window updates will only happen at the same time when there's only one stream. With multiple streams, the windows get out of sync. Once we introduce a different window size for the connection vs the streams, the windows won't even be in sync in the one stream scenario.

**Comment by halter73** [comment]
⬆️📅

**Comment by halter73** [comment]
@dotnet-bot test OSX 10.12 Release Build

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs [hunk 3]
```diff
@@ -23,24 +21,35 @@ public partial class Http2Stream : HttpProtocol
...
             Output = _http2Output;
         }
 
         public int StreamId => _context.StreamId;
 
         public bool RequestBodyStarted { get; private set; }
-        public bool EndStreamReceived { get; private set; }
-
-        protected IHttp2StreamLifetimeHandler StreamLifetimeHandler => _context.StreamLifetimeHandler;
+        public bool EndStreamReceived => (_completionState & StreamCompletionFlags.EndStreamReceived) == StreamCompletionFlags.EndStreamReceived;
```
Comment (Tratcher) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: I see what you meant by wanting to use HasFlag

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs [hunk 4]
```diff
@@ -315,6 +360,60 @@ private void AbortCore(ConnectionAbortedException abortReason)
...
+            ));
+
+        private bool TryApplyCompletionFlag(StreamCompletionFlags completionState)
+        {
+            lock (_completionLock)
+            {
+                var lastCompletionState = _completionState;
+                _completionState |= completionState;
+
+                if (ShoulStopTrackingStream(_completionState) && !ShoulStopTrackingStream(lastCompletionState))
```
Comment (benaadams) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: Shoul**d**StopTrackingStream

Comment (benaadams) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: Missing d
ShoulStopTrackingStream
ShouldStopTrackingStream

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs (hunk 3 above)
Comment (benaadams) [reply src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: What's wrong with HasFlag? Is an intrinsic now https://github.com/dotnet/coreclr/pull/13748

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: The method exists in desktop though? Its just slower?

Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: Yep. The desktop version of HasFlag does reflection each time AFAIK.

**Pull request merged**